### PR TITLE
Add support for postgres db backend

### PR DESCRIPTION
**Pull request by tempoz**
This PR actually adds support for the postgres backend via the gorm postgres
driver.

**Related issues**: N/A

### CODE REVIEW

**Comment by tempoz** [comment]
Current dependencies on/for this PR:
* master
  * **PR #3821** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/3821" target="_blank">Graphite</a> 
    * **PR #3822** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/3822" target="_blank">Graphite</a> 
      * **PR #3823** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/3823" target="_blank">Graphite</a> 
        * **PR #3824** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/3824" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/3824?utm_source=stack-comment).

**Comment by tempoz** [comment]
Current dependencies on/for this PR:
* master
  * **PR #3821** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/3821" target="_blank">Graphite</a> 
    * **PR #3822** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/3822" target="_blank">Graphite</a> 
      * **PR #3823** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/3823" target="_blank">Graphite</a> 
        * **PR #3824** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/3824" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/3824?utm_source=stack-comment).

**Review by bduffany — COMMENTED** [review]
not sure if you are already thinking about this, but I think it would be great to have an automated test on CI for this since we won't be actively exercising the postgres support ourselves (for now?)

You can take a look at how `usage_test_mysql` is implemented and see how we run a mysql server using docker (testmysql.go) - we could probably do something pretty similar for postgres, and then just have a small test that runs the GORM auto-migrate logic against an empty postgres DB.

The test would need to have the same `tags` and `exec_properties` as usage_test_mysql to make sure it runs using docker-in-firecracker on CI

File: server/testutil/testpostgres/testpostgres.go
Comment (bduffany) [inline server/testutil/testpostgres/testpostgres.go:105]: nit: uncomment?

**Review by bduffany — APPROVED** [review]
LGTM, nice :)

would also be good to get a second review from Vadim

File: server/tables/tables.go
Comment (vadimberezniker) [reply server/tables/tables.go:901]: +1 don't need this migration for postgres here since we'll never have data w/o the uuid filled

File: server/util/db/db.go
Comment (vadimberezniker) [inline server/util/db/db.go:287]: It looks like the postgres driver supports the connection string in the same format as MySQL and sqlite: https://pkg.go.dev/github.com/jackc/pgx/v5/pgconn#ParseConfig

Would it be simpler if we stuck to this format instead of using something different for postgres?

**Review by vadimberezniker — CHANGES_REQUESTED** [review]

File: enterprise/server/usage/BUILD
Comment (bduffany) [inline enterprise/server/usage/BUILD:92]: I realized that we could have only the test action run with docker-in-firecracker + runner recycling by applying a `test.` prefix to these. I made the changes in https://github.com/buildbuddy-io/buildbuddy/pull/4055 - mind applying this diff here as well?

```suggestion
        "test.workload-isolation-type": "firecracker",
        "test.init-dockerd": "true",
        "test.recycle-runner": "true",
```

File: enterprise/server/backends/userdb/userdb.go
Comment (vadimberezniker) [inline enterprise/server/backends/userdb/userdb.go:786]: Does postgres not like the query w/o this?

File: enterprise/server/usage/usage.go
Comment (vadimberezniker) [inline enterprise/server/usage/usage.go:298]: Was this necessary because the previous query did not work for postgres?

File: enterprise/server/backends/userdb/userdb.go
Comment (vadimberezniker) [reply enterprise/server/backends/userdb/userdb.go:786]: Assuming the answer is yes, I would suggest adding an explicit `ON` clause rather than relying on implicit column matching.

File: enterprise/server/usage/usage.go
Comment (tempoz) [reply enterprise/server/usage/usage.go:298]: That's correct, postgres has a very different syntax for ignoring conflicts.

File: enterprise/server/backends/userdb/userdb.go
Comment (tempoz) [reply enterprise/server/backends/userdb/userdb.go:786]: It does not; I can use an `ON` clause.

File: enterprise/server/usage/usage.go
Comment (tempoz) [reply enterprise/server/usage/usage.go]: Oh, you're right, I'm sorry. This change did end up not being necessary and this state is a result of iterative development when fixing this transaction as a whole; I can revert this query.

File: server/tables/tables.go
Comment (vadimberezniker) [inline server/tables/tables.go:873]: this function is only called for mysql so there's no need for this switch

File: server/util/db/db.go
Comment (vadimberezniker) [inline server/util/db/db.go:581]: What's the reason for removing this check?

**Review by vadimberezniker — APPROVED** [review]

File: server/util/db/db.go
Comment (tempoz) [reply server/util/db/db.go:581]: I think it got moved and then removed from the place it was moved to; I will reinstate it.

File: enterprise/server/backends/userdb/userdb.go
Comment (tempoz) [reply enterprise/server/backends/userdb/userdb.go:786]: Sure, I'll remove it.

**Pull request merged**
